zrml/authorized/src/lib.rs

```diff
@@ -142,23 +193,60 @@ mod pallet {
         type Origin = T::Origin;
 
         fn on_dispute(
-            _: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
+            disputes: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
             _: &Self::MarketId,
-            _: &Market<Self::AccountId, Self::BlockNumber, Self::Moment>,
+            market: &Market<Self::AccountId, Self::BlockNumber, Self::Moment>,
         ) -> DispatchResult {
+            ensure!(
+                market.dispute_mechanism == MarketDisputeMechanism::Authorized,
+                Error::<T>::MarketDoesNotHaveDisputeMechanismAuthorized
+            );
+            ensure!(disputes.is_empty(), Error::<T>::OnlyOneDisputeAllowed);
             Ok(())
         }
 
         fn on_resolution(
             _: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
             market_id: &Self::MarketId,
-            _: &Market<Self::AccountId, Self::BlockNumber, MomentOf<T>>,
+            market: &Market<Self::AccountId, Self::BlockNumber, MomentOf<T>>,
         ) -> Result<Option<OutcomeReport>, DispatchError> {
+            ensure!(
+                market.dispute_mechanism == MarketDisputeMechanism::Authorized,
+                Error::<T>::MarketDoesNotHaveDisputeMechanismAuthorized
+            );
             let result = AuthorizedOutcomeReports::<T>::get(market_id);
             if result.is_some() {
                 AuthorizedOutcomeReports::<T>::remove(market_id);
             }
-            Ok(result)
+            Ok(result.map(|report| report.outcome))
+        }
+
+        fn get_auto_resolve(
+            _: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
+            market_id: &Self::MarketId,
+            market: &Market<Self::AccountId, Self::BlockNumber, MomentOf<T>>,
+        ) -> Result<Option<Self::BlockNumber>, DispatchError> {
+            ensure!(
+                market.dispute_mechanism == MarketDisputeMechanism::Authorized,
+                Error::<T>::MarketDoesNotHaveDisputeMechanismAuthorized
+            );
+            Ok(Self::get_auto_resolve(market_id))
+        }
+
+        fn is_fail(
+            _: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
+            market_id: &Self::MarketId,
+            market: &Market<Self::AccountId, Self::BlockNumber, MomentOf<T>>,
+        ) -> Result<bool, DispatchError> {
+            ensure!(
+                market.dispute_mechanism == MarketDisputeMechanism::Authorized,
+                Error::<T>::MarketDoesNotHaveDisputeMechanismAuthorized
+            );
+            let is_unreported = !AuthorizedOutcomeReports::<T>::contains_key(market_id);
+            let report = market.report.as_ref().ok_or(Error::<T>::MarketIsNotReported)?;
+            let now = frame_system::Pallet::<T>::block_number();
+            let is_expired = report.at.saturating_add(T::ReportPeriod::get()) < now;
```

Comment:
> we sometimes have to delay the resolution of a market. [...] but the correct outcome will only be clear a month later.

This example is really a good point. How do we would like to handle something like this in `court`? I mean shouldn't be there some option, to delay a resolution of a market? This seems like a hacky solution, although it works perfectly for `authorized`. I mean, when the outcome is not clear, the market should have a resolution delay.

"migration-safe"

True, I could save the end time in storage rather than using a config parameter. I would, because I think this is a great argument. But when we decide to remove the optionality to resolve a market after a specific report period...

Let me think about this...

Reply:
Ok, got you. But I still prepared our system for the account id usage as the authority.

0f3b1e12d4d4a973016ddee576a2a086192b6dcd

Reply:
How do we would like to handle something like this in court? I mean shouldn't be there some option, to delay a resolution of a market?

We have an issue for this already: #762. Well, the option is already there: As long as the dispute system doesn't queue an auto-resolution, the resolution is delayed. From here it's a matter of implementation. The authority can just wait until the result of the event becomes apparent (although for an account-based authority, we still need a solution!). Not sure what a good solution for court is. We should check the Kleros implementation. Maybe we should add an outcome for delaying resolution, but I'm not sure what the decision theoretic implications are. Needs more research.